lib/srv/db/mysql/gcp.go

```diff
@@ -0,0 +1,138 @@
+)
+
+func isDBUserGCPServiceAccount(dbUser string) bool {
+	if strings.Contains(dbUser, "@") {
+		switch {
+		// Example: mysql-iam-user@my-project-id.iam
+		// This format is used to align with PostgreSQL.
+		case strings.HasSuffix(dbUser, ".iam"):
+			return true
+		// Example: mysql-iam-user@my-project-id..gserviceaccount.com
+		case strings.HasSuffix(dbUser, ".iam.gserviceaccount.com"):
+			return true
+		}
+	}
+	return false
+}
+
+func gcpServiceAccountToDatabaseUser(serviceAccountName string) string {
+	user, _, _ := strings.Cut(serviceAccountName, "@")
+	return user
+}
+
+func databaseUserToGCPServiceAccount(sessionCtx *common.Session) string {
+	return fmt.Sprintf("%s@%s.iam.gserviceaccount.com", sessionCtx.DatabaseUser, sessionCtx.Database.GetGCP().ProjectID)
+}
+
+func (e *Engine) getGCPUserAndPassword(ctx context.Context, sessionCtx *common.Session, gcpClient gcp.SQLAdminClient) (string, string, error) {
+	// If `--db-user` is an service account email, use IAM Auth.
+	if isDBUserGCPServiceAccount(sessionCtx.DatabaseUser) {
+		return e.getGCPIAMUserAndPassword(ctx, sessionCtx)
+	}
+
+	// Get user info to decide how to authenticate.
+	user, err := gcpClient.GetUser(ctx, sessionCtx.Database, sessionCtx.DatabaseUser)
+	if err != nil {
+		// GetUser permission is new for IAM auth. If no permission, assume legacy password user.
+		if trace.IsAccessDenied(err) {
+			return e.getGCPPasswordUserAndPassword(ctx, sessionCtx)
+		}
+		return "", "", trace.Wrap(err)
+	}
+
+	// Possible values (copied from SDK):
+	//   "BUILT_IN" - The database's built-in user type.
+	//   "CLOUD_IAM_USER" - Cloud IAM user.
+	//   "CLOUD_IAM_SERVICE_ACCOUNT" - Cloud IAM service account.
+	//   "CLOUD_IAM_GROUP" - Cloud IAM group non-login user.
+	//   "CLOUD_IAM_GROUP_USER" - Cloud IAM group login user.
+	//   "CLOUD_IAM_GROUP_SERVICE_ACCOUNT" - Cloud IAM group service account.
+	//
+	// In practice, type can also be empty for built-in user.
+	switch user.Type {
+	case "",
+		"BUILT_IN":
+		return e.getGCPPasswordUserAndPassword(ctx, sessionCtx)
+
+	case "CLOUD_IAM_SERVICE_ACCOUNT",
+		"CLOUD_IAM_GROUP_SERVICE_ACCOUNT":
+		return e.getGCPIAMUserAndPassword(
+			ctx,
+			sessionCtx.WithUser(databaseUserToGCPServiceAccount(sessionCtx)),
+		)
+
+	default:
+		return "", "", trace.BadParameter("GCP MySQL user type %q not supported", user.Type)
+	}
+}
+
+func (e *Engine) getGCPIAMUserAndPassword(ctx context.Context, sessionCtx *common.Session) (string, string, error) {
+	e.Log.WithField("session", sessionCtx).Debug("Authenticating GCP MySQL with IAM auth.")
+
+	// Note that sessionCtx.DatabaseUser is the service account.
+	password, err := e.Auth.GetCloudSQLAuthToken(ctx, sessionCtx)
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+	return gcpServiceAccountToDatabaseUser(sessionCtx.DatabaseUser), password, nil
+}
+
+func (e *Engine) getGCPPasswordUserAndPassword(ctx context.Context, sessionCtx *common.Session) (string, string, error) {
```

Comment:
The function name is a bit uncer for me -  get*Password*AndPassword

What about getGCPPasswordAuth()

Do we need to return the sessionCtx.DatabaseUser ? Note that this is part of the input param sessionCtx and the DatabaseUser prope is not modified in this function.

Reply:
Renamed to `getGCPIAMAuthToken` and `getGCPOneTimePassword`, and removed the user name from the return.